### ISSUE TITLE
Consider adding a telemetry event for failed HTTP connection attempts

### ISSUE THREAD

**Issue by MihaZupan** [dotnet/runtime#110351]
With how requests are split from connections in `SocketsHttpHandler`, the user may not always see the errors that occurred during connection establishment (dns + tcp + tls + http/2 handshake).
The errors may be swallowed if we already server the initiating request with a different connection, or if that request timed out.
There are therefore situations where you may have poor visibility into what's happening.

One way of gaining the info is to inject a custom `ConnectCallback` along the lines of
```c#
handler.ConnectCallback = async (context, ct) =>
{
    var stopwatch = Stopwatch.StartNew();
    var socket = new Socket(SocketType.Stream, ProtocolType.Tcp) { NoDelay = true };
    try
    {
        await socket.ConnectAsync(context.DnsEndPoint, ct);
        return new NetworkStream(socket, ownsSocket: true);
    }
    catch (Exception ex)
    {
        socket.Dispose();
        Console.WriteLine($"Failed to connect to {context.DnsEndPoint} after {stopwatch.ElapsedMilliseconds:N2}: {ex}");
        throw;
    }
};
```
This gives you visibility into failures when trying to connect, but not into TLS failures/timeouts.
It's possible to also perform the TLS handshake in the callback, but we don't always make it trivial (how do you pick the host, how do you handle ALPN, potential custom ssl settings, ...).

---

It seems rather simple for us to add an `EventSource` event to `System.Net.Http` that just logs all such failures, corresponding to
https://github.com/dotnet/runtime/blob/05fa8813a7774890b46e60109d8cd6d4c853a014/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http1.cs#L307-L309
https://github.com/dotnet/runtime/blob/05fa8813a7774890b46e60109d8cd6d4c853a014/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http2.cs#L271-L273
https://github.com/dotnet/runtime/blob/05fa8813a7774890b46e60109d8cd6d4c853a014/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http3.cs#L325-L329

Maybe something like
```c#
[Event(42, Level = EventLevel.Error)]
public void ConnectionFailed(byte versionMajor, byte versionMinor, string scheme, string host, int port, string? remoteAddress, double elapsedMilliseconds, string exception);
```

Does this seem reasonable @antonfirsov, maybe we have better ways of exposing such data now?

**Comment by antonfirsov** [comment]
> maybe we have better ways of exposing such data now?

We are reporting the same exceptions on the `HTTP connection_setup` Activity when a connection attempt fails:

https://github.com/dotnet/runtime/blob/efd5a63ac0681203c1a05e667669e145a188182d/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.cs#L647-L653

Whether it's a "better way", depends on the use-case. Some may prefer to consume distributed traces, others may want `EventSource` logs. IMO it would be valuable to expose the information in both telemetry legs, but the information should be logged at the same point, and I would prefer it to happen around the line I linked. The difference is that at that point we don't catch exceptions comin from  `PlainTextStreamFilter` and cancellations are not wrapped yet by `CreateConnectTimeoutException`.

**Comment by MihaZupan** [comment]
Ah, thanks, I forgot about this one. An event could still be useful but I'm not as worried in that case.

> are not wrapped yet by CreateConnectTimeoutException.

Looks like we're inconsistent here rn, we do report the ConnectTimeout on H/3.

https://github.com/dotnet/runtime/blob/05fa8813a7774890b46e60109d8cd6d4c853a014/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http3.cs#L280-L289

I could see it being useful if we did include info on whether the cancellation was due to ConnectTimeout / due to `CancelledByOriginatingRequestCompletion` (we don't include that info today anyway, but we could improve that).

One difference that could matter is that we also exclude H/2 handshake failures here.

**Comment by antonfirsov** [comment]
> An event could still be useful but I'm not as worried in that case.

So are you aware of a specific user scenario you wanted this to help with? If yes, does this mean they would be OK consuming distributed traces?

**Comment by MihaZupan** [comment]
Yes. I'm helping in an investigation where a service isn't establishing connections _for some reason_.
In trying to advise what sort of logging might help us investigate the issue (outside of internal diagnostics), but I can't give a good, clear, simple answer atm. The `ConnectCallback` can cover some, but not all, cases.

> does this mean they would be OK consuming distributed traces?

I'll try to find out, but it's likely that it only being available since 9.0 is going to be an issue.

**Comment by CarnaViire** [comment]
Triage: doesn't seem to be critical for 10.0, punting to 11.0

P.S.: @antonfirsov @MihaZupan could you pls clarify what was agreed on for the scope here? I.e. should improving/unifying exception behavior be included in this issue, or tracked separately?
